From 536f80376d28f79c8ebb8c373ead89bff98bc7ce Mon Sep 17 00:00:00 2001
From: Raphael Mounier <mounierr07@gmail.com>
Date: Thu, 1 Jun 2023 15:54:24 +0200
Subject: [PATCH] core :  add led support for offline charger

Just allows you to change the color of the LEDs depending on the state of charge of the battery

Change-Id: Ieac086cf1cb5521a28af7a42a09ff41ef89d3cf4
---
 healthd/healthd_draw.cpp                      |  8 ++--
 healthd/healthd_mode_charger.cpp              | 37 ++++++++++++++++---
 .../charger/healthd_mode_charger.h            |  3 ++
 3 files changed, 38 insertions(+), 10 deletions(-)

diff --git a/healthd/healthd_draw.cpp b/healthd/healthd_draw.cpp
index 2266233e1..f469dff20 100644
--- a/healthd/healthd_draw.cpp
+++ b/healthd/healthd_draw.cpp
@@ -97,7 +97,7 @@ HealthdDraw::HealthdDraw(animation* anim)
         if (!strcmp(prop_hardware,"hi3660")
             || !strcmp(prop_hardware,"hi3670")
             || !strcmp(prop_hardware,"hi6250")
-            || !strcmp(prop_hardware,"kirin"))
+            || !strncmp(prop_hardware,"kirin",5))
         {
             LOGV("Kirin Huawei found\n");
             is_kirin=true;
@@ -139,7 +139,7 @@ void HealthdDraw::set_brightness(uint32_t value) {
     LOGV("Kirin - Try to set brightness to %d\n",value)
     if (WriteStringToFile(std::to_string(value), "/sys/class/leds/lcd_backlight0/brightness")==false) {
         LOGW("Kirin - WriteStringToFile failed lcd_backlight0, unable to set brightness (lcd_backlight0)\n");
-        if (WriteStringToFile(std::to_string(0), "/sys/class/leds/lcd_backlight/brightness")==false) {
+        if (WriteStringToFile(std::to_string(value), "/sys/class/leds/lcd_backlight/brightness")==false) {
             LOGE("Kirin - WriteStringToFile failed lcd_backlight, unable to set brightness (lcd_backlight)\n");
         }
     }
@@ -159,8 +159,8 @@ void HealthdDraw::blank_screen(bool blank, int drm) {
     if (is_kirin) {
         if (blank==true) {
             LOGV("Kirin - clear screen\n")
-            //clear_screen();
-            //gr_flip();
+            clear_screen();
+            gr_flip();
             set_brightness(0);
         }
         else {
diff --git a/healthd/healthd_mode_charger.cpp b/healthd/healthd_mode_charger.cpp
index f01ffd486..1ca573cbd 100644
--- a/healthd/healthd_mode_charger.cpp
+++ b/healthd/healthd_mode_charger.cpp
@@ -64,6 +64,7 @@ using std::string_literals::operator""s;
 using namespace android;
 using aidl::android::hardware::health::BatteryStatus;
 using android::hardware::health::HealthLoop;
+using android::base::WriteStringToFile;
 
 // main healthd loop
 extern int healthd_main(void);
@@ -306,15 +307,39 @@ void Charger::UpdateLedState() {
     if (!have_battery_state_) return;
     if (health_info_.battery_level == 0 && health_info_.battery_status == BatteryStatus::UNKNOWN) return ;
 
-    // TODO set led with battery_level in % (0->100%)
     LOGV("Battery level = %d\n",health_info_.battery_level);
 
-    /*
-    /sys/class/leds/red/brightness
-    /sys/class/leds/green/brightness
-    /sys/class/leds/blue/brightness
-    */
+    // Calculate new led color
+    uint8_t red=0;
+    uint8_t green=0;
+    uint8_t blue=0;
 
+    if (health_info_.battery_level>=0 && health_info_.battery_level<10) {
+        red=255;green=0;
+    }
+	if (health_info_.battery_level>=21 && health_info_.battery_level<89) {
+        red=128;green=128;
+    }
+	if (health_info_.battery_level>=90 && health_info_.battery_level<=100) {
+        red=0;green=255;
+    }
+
+    if ((red==old_red) && (green==old_green))
+       return;
+
+    // change led color
+    old_red=red;
+    old_green=green;
+    LOGV("Kirin - WriteStringToFile Leds Color : red %d, green %d\n",red,green);
+    if (WriteStringToFile(std::to_string(red), "/sys/class/leds/red/brightness")==false) {
+        LOGE("Kirin - WriteStringToFile failed, unable to set red led\n");
+    }
+    if (WriteStringToFile(std::to_string(green), "/sys/class/leds/green/brightness")==false) {
+        LOGE("Kirin - WriteStringToFile failed, unable to set green led\n");
+    }
+    if (WriteStringToFile(std::to_string(blue), "/sys/class/leds/blue/brightness")==false) {
+        LOGE("Kirin - WriteStringToFile failed, unable to set blue led\n");
+    }
 }
 
 
diff --git a/healthd/include_charger/charger/healthd_mode_charger.h b/healthd/include_charger/charger/healthd_mode_charger.h
index 291fc7eb3..9d8bb5498 100644
--- a/healthd/include_charger/charger/healthd_mode_charger.h
+++ b/healthd/include_charger/charger/healthd_mode_charger.h
@@ -117,6 +117,9 @@ class Charger {
     int64_t next_key_check_ = 0;
     int64_t next_pwr_check_ = 0;
     int64_t wait_batt_level_timestamp_ = 0;
+    
+    uint8_t old_red = 0;
+    uint8_t old_green = 0;
 
     DirectRenderManager drm_;
     SrceenSwitch screen_switch_;
-- 
2.25.1

